tests/keras_autolog/test_keras_autolog.py

```diff
@@ -149,8 +152,148 @@ def test_keras_autolog_logs_default_params(keras_random_data_run):
+
+@pytest.fixture
+def all_zeros_train_data():
+    return np.zeros((1000, 32))
+
+
+@pytest.fixture
+def all_zeros_one_hot_labels():
+    n, n_class = (1000, 10)
+    classes = np.zeros(n, dtype=int)
+    labels = np.zeros((n, n_class))
+    labels[np.arange(n), classes] = 0
+    return labels
+
+
+@pytest.fixture
+def keras_random_data_run_with_callback(all_zeros_train_data, fit_variant,
+                                        all_zeros_one_hot_labels, manual_run,
+                                        callback, restore_weights, patience):
+    mlflow.keras.autolog()
+
+    data = all_zeros_train_data
+    labels = all_zeros_one_hot_labels
+
+    model = create_model()
+    if callback == 'early':
+        callback = keras.callbacks.callbacks.EarlyStopping(monitor='loss', patience=patience,
+                                                           restore_best_weights=restore_weights)
+    else:
+        if fit_variant == 'fit_generator':
+            count_mode = 'steps'
+        else:
+            count_mode = 'samples'
+        callback = keras.callbacks.callbacks.ProgbarLogger(count_mode=count_mode)
+
+    if fit_variant == 'fit_generator':
+        def generator():
+            while True:
+                yield data, labels
+        history = model.fit_generator(generator(), epochs=10, callbacks=[callback],
+                                      steps_per_epoch=1, shuffle=False)
+    else:
+        history = model.fit(data, labels, epochs=10, callbacks=[callback])
+
+    client = mlflow.tracking.MlflowClient()
+    return client.get_run(client.list_run_infos(experiment_id='0')[0].run_id), history, callback
+
+
+@pytest.mark.large
+@pytest.mark.parametrize('fit_variant,restore_weights', [('fit', True),
+                                                         ('fit_generator', True)])
+@pytest.mark.parametrize('callback', ['early'])
+@pytest.mark.parametrize('patience', [0, 1, 5])
+def test_keras_autolog_early_stop_logs(keras_random_data_run_with_callback):
+    run, history, callback = keras_random_data_run_with_callback
+    metrics = run.data.metrics
+    params = run.data.params
+    assert 'stopped_epoch' in metrics
+    assert 'earlystopping_patience' in params
+    assert params['earlystopping_patience'] == str(callback.patience)
+    assert 'best_epoch' in metrics
+    assert int(metrics['stopped_epoch']) - max(1, callback.patience) == int(metrics['best_epoch'])
+    assert 'loss' in history.history
+    num_of_epochs = len(history.history['loss'])
+    client = mlflow.tracking.MlflowClient()
+    metric_history = client.get_metric_history(run.info.run_id, 'loss')
+    # Check the test epoch numbers are correct
+    assert num_of_epochs == max(1, callback.patience) + 1
+    # Check that MLflow has logged the metrics of the "best" model
+    assert len(metric_history) == num_of_epochs + 1
+    # Check that MLflow has logged the correct data
+    assert history.history['loss'][callback.stopped_epoch] == metric_history[-1].value
```

Comment:
Is this line correct? If early stopping is triggered, metrics at the stopped epoch is NOT the best metrics.

Reply:
Yes, this is incorrect. The test itself wasn't robust - I went ahead and fixed this/refactored the tests so that it would work correctly. Thanks for the catch!